### PR TITLE
chore: e2e test cases for `showkeys` and `showcommands'

### PR DESCRIPTION
**Pull request by rishabhKalakoti**
## Description

The following e2e tests cases were written for the command `teachcode showkeys`:

- [x] No config file present
- [x] No keys in the config file
- [x] Multiple keys in the config file

The following e2e tests cases were written for the command `teachcode showcommands`:

- [x] No config file present
- [x] Valid config file present

addresses #28 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -11,13 +11,16 @@
     "teachcode": "./bin/index.js"
   },
   "scripts": {
-    "lint": "eslint src bin test",
-    "lint:fix": "eslint --fix src bin test",
+    "lint": "eslint src bin __e2e__",
+    "lint:fix": "eslint --fix src bin __e2e__",
     "pretest": "npm run lint",
-    "test": "ava test/index.js",
+    "test": "ava --serial",
```
Comment (rishabhKalakoti) [inline package.json]: `ava` tests run serially as tests from different files may alter the same `config.json` file from the filesystem.

File: __e2e__/commands/showcommands.test.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    reject: false,
+  });
+  t.snapshot(stdout);
+});
+
+test.serial('config file present for showcommands', async t => {
+  fs.writeFileSync('config.json', JSON.stringify(validConfigJson));
+  const { stdout } = await execa(rootCommand, ['showcommands'], {
+    reject: false,
+  });
```
Comment (jamesgeorge007) [inline __e2e__/commands/showcommands.test.js]:
```js
const { stdout } = await execa(rootCommand, ['showcommands']);
```
`Promise` is not rejected.

File: __e2e__/commands/showcommands.test.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+  t.snapshot(stdout);
+  await execa('rm', ['config.json']);
```
Comment (jamesgeorge007) [inline __e2e__/commands/showcommands.test.js]: Not a cross-platform solution. Go with [rimraf](https://github.com/isaacs/rimraf) instead.

File: __e2e__/commands/showkeys.test.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+const path = require('path');
+const execa = require('execa');
+const test = require('ava');
+const fs = require('fs');
+
+const rootCommand = path.join(process.cwd(), 'bin/index.js');
+const configFilePath = path.join(process.cwd(), 'config.json');
+
+var jsonNoKeys = { userName: 'TestJsonNoKeys', taskCount: 0, keys: [] };
```
Comment (jamesgeorge007) [inline __e2e__/commands/showkeys.test.js]:
```suggestion
const configWithoutKeys = { userName: 'TestJsonNoKeys', taskCount: 0, keys: [] };
```

File: __e2e__/commands/showkeys.test.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+var jsonMultipleKeys = {
```
Comment (jamesgeorge007) [inline __e2e__/commands/showkeys.test.js]:
```suggestion
const configWithMultipleKeys = {
```

File: __e2e__/commands/showkeys.test.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+
+test.serial('no keys in config file', async t => {
+  fs.writeFileSync('config.json', JSON.stringify(jsonNoKeys));
+  const { stdout } = await execa(rootCommand, ['showkeys'], { reject: false });
+  t.snapshot(stdout);
+  await execa('rm', ['config.json']);
+});
+
+test.serial('multiple keys in config file', async t => {
+  fs.writeFileSync('config.json', JSON.stringify(jsonMultipleKeys));
```
Comment (jamesgeorge007) [inline __e2e__/commands/showkeys.test.js]:
```suggestion
  fs.writeFileSync('config.json', JSON.stringify(configWithMultipleKeys));
```

File: __e2e__/commands/showkeys.test.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+// tests are serial as we are creating different temp config files for each tests
+test.serial('no config file', async t => {
+  if (fs.existsSync(configFilePath)) await execa('rm', ['config.json']);
+  const { stdout } = await execa(rootCommand, ['showkeys'], { reject: false });
+  t.snapshot(stdout);
+});
+
+test.serial('no keys in config file', async t => {
+  fs.writeFileSync('config.json', JSON.stringify(jsonNoKeys));
```
Comment (jamesgeorge007) [inline __e2e__/commands/showkeys.test.js]:
```suggestion
  fs.writeFileSync('config.json', JSON.stringify(configWithoutKeys));
```

File: package.json [hunk 8]
```diff
@@ -11,13 +11,16 @@
...
     "docs:dev": "vuepress dev docs",
     "docs:build": "vuepress build docs"
   },
+  "ava": {
+    "snapshotDir": "__snapshots__"
```
Comment (jamesgeorge007) [inline package.json]:
```suggestion
    "snapshotDir": "__e2e__/__snapshots__"
```

File: __e2e__/commands/showkeys.test.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  const { stdout } = await execa(rootCommand, ['showkeys'], { reject: false });
```
Comment (jamesgeorge007) [inline __e2e__/commands/showkeys.test.js]: https://github.com/madlabsinc/teachcode/pull/63/files#r361797335

File: __e2e__/commands/showkeys.test.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  t.snapshot(stdout);
+  await execa('rm', ['config.json']);
+});
+
+test.serial('multiple keys in config file', async t => {
+  fs.writeFileSync('config.json', JSON.stringify(jsonMultipleKeys));
+  const { stdout } = await execa(rootCommand, ['showkeys'], { reject: false });
```
Comment (jamesgeorge007) [inline __e2e__/commands/showkeys.test.js]: https://github.com/madlabsinc/teachcode/pull/63/files#r361797335

File: __e2e__/commands/showkeys.test.js [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+  t.snapshot(stdout);
+  await execa('rm', ['config.json']);
```
Comment (jamesgeorge007) [inline __e2e__/commands/showkeys.test.js]: https://github.com/madlabsinc/teachcode/pull/63/files#r361797423

File: __e2e__/commands/showkeys.test.js [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+  if (fs.existsSync(configFilePath)) await execa('rm', ['config.json']);
+  const { stdout } = await execa(rootCommand, ['showkeys'], { reject: false });
+  t.snapshot(stdout);
+});
+
+test.serial('no keys in config file', async t => {
+  fs.writeFileSync('config.json', JSON.stringify(jsonNoKeys));
+  const { stdout } = await execa(rootCommand, ['showkeys'], { reject: false });
+  t.snapshot(stdout);
+  await execa('rm', ['config.json']);
```
Comment (jamesgeorge007) [inline __e2e__/commands/showkeys.test.js]: https://github.com/madlabsinc/teachcode/pull/63/files#r361797423

File: __e2e__/commands/showkeys.test.js [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+var jsonNoKeys = { userName: 'TestJsonNoKeys', taskCount: 0, keys: [] };
+var jsonMultipleKeys = {
+  userName: 'TestJsonMultipleKeys',
+  taskCount: 2,
+  keys: ['testkey1', 'testkey2'],
+};
+
+// tests are serial as we are creating different temp config files for each tests
+test.serial('no config file', async t => {
+  if (fs.existsSync(configFilePath)) await execa('rm', ['config.json']);
```
Comment (jamesgeorge007) [inline __e2e__/commands/showkeys.test.js]: https://github.com/madlabsinc/teachcode/pull/63/files#r361797423

File: __e2e__/commands/showcommands.test.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+const rootCommand = path.join(process.cwd(), 'bin/index.js');
+const configFilePath = path.join(process.cwd(), 'config.json');
+
+const validConfigJson = {
+  userName: 'validConfigJson',
+  taskCount: 0,
+};
+
+test.serial('no config file for showcommands', async t => {
+  if (fs.existsSync(configFilePath)) await execa('rm', ['config.json']);
```
Comment (jamesgeorge007) [inline __e2e__/commands/showcommands.test.js]: https://github.com/madlabsinc/teachcode/pull/63/files#r361797423

File: package.json (hunk 8 above)
Comment (rishabhKalakoti) [reply package.json]: Doing this gives the following directory structure:
--- \_\_e2e\_\_/
    |---------cli.test.js
    |---------commands/
    |---------\_\_snapshots\_\_/
        |------------\_\_e2e\_\_/
            |--------------cli.test.js.md
            |--------------cli.test.js.snap
            |--------------commands/
                     |-----------------showkeys.test.js.md
                     |-----------------showkeys.test.js.snap
                     |-----------------showcommands.test.js.md
                     |-----------------showcommands.test.js.snap

But I don't think this is what we desire.

File: __e2e__/commands/showcommands.test.js (hunk 3 above)
Comment (rishabhKalakoti) [reply __e2e__/commands/showcommands.test.js]: Is it fine to go with [fs unlink sync](https://nodejs.org/api/fs.html#fs_fs_unlinksync_path)? It seems better for promises

Comment (jamesgeorge007) [reply __e2e__/commands/showcommands.test.js]: Yeah, let's not introduce new dependencies.

File: package.json (hunk 8 above)
Comment (jamesgeorge007) [reply package.json]: Everything related to `e2e tests` should come under the `__e2e__` directory.

File: __e2e__/commands/showcommands.test.js (hunk 3 above)
Comment (rishabhKalakoti) [reply __e2e__/commands/showcommands.test.js]: Not sure if I should use [fs unlink](https://nodejs.org/api/fs.html#fs_fs_unlink_path_callback) instead. The tests are passing with `fs.unlinkSync` though.

**Pull request merged**